From 3f6a8fb3bdafdf3ae08cf1416fe13a5b0a604a68 Mon Sep 17 00:00:00 2001
From: Suraj P Kizhakkethil <quic_surapk@quicinc.com>
Date: Sat, 28 Dec 2024 17:57:02 +0530
Subject: [PATCH] QSDK:CP:wifi-scripts: Add support for mac address generation
 for mbssid

Currently, after openwrt upgrade, the mac address generation for mbssid
is incorrect and not as per 802.11 specification. Add support to generate
mac addresses for mbssid as per 802.11 specification.

Change-Id: I50f48dbb1b1d8b10deceaca8f522781f1acbc273
Co-developed-by: Harshitha Prem <quic_hprem@quicinc.com>
Signed-off-by: Harshitha Prem <quic_hprem@quicinc.com>
Signed-off-by: Suraj P Kizhakkethil <quic_surapk@quicinc.com>
---
 .../files/lib/netifd/wireless/mac80211.sh     |  8 +++++-
 .../files/usr/share/hostap/common.uc          | 26 +++++++++++++++++--
 2 files changed, 31 insertions(+), 3 deletions(-)

diff --git a/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh b/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
index fb17097f8a32..af2c88c2f0b9 100755
--- a/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
+++ b/package/network/config/wifi-scripts/files/lib/netifd/wireless/mac80211.sh
@@ -924,8 +924,14 @@ mac80211_get_addr() {
 mac80211_generate_mac() {
 	local phy="$1"
 	local id="${macidx:-0}"
+	local group_size
+	if [ -n "mbssid_group_size" ]; then
+		group_size="$mbssid_group_size"
+	else
+		group_size="$has_ap"
+	fi
 
-	wdev_tool "$phy$phy_suffix" get_macaddr id=$id num_global=$num_global_macaddr mbssid=${multiple_bssid:-0}
+	wdev_tool "$phy$phy_suffix" get_macaddr id=$id num_global=$num_global_macaddr mbssid=$multiple_bssid mbssid_group_size=$group_size
 }
 
 get_board_phy_name() (
diff --git a/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc b/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc
index c6bfb62ef3c5..81ba48bd4c11 100644
--- a/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc
+++ b/package/network/config/wifi-scripts/files/usr/share/hostap/common.uc
@@ -204,6 +204,7 @@ const phy_proto = {
 		let mbssid = int(data.mbssid ?? 0) > 0;
 		let num_global = int(data.num_global ?? 1);
 		let use_global = !mbssid && idx < num_global;
+		let mbssid_group_size = int(data.mbssid_group_size ?? 4);
 
 		let base_addr = phy_sysfs_file(phy, "macaddress");
 		if (!base_addr)
@@ -234,6 +235,21 @@ const phy_proto = {
 			return base_addr;
 
 		let addr = macaddr_split(base_addr);
+		let maxbssid_ind = 0;
+		if (mbssid) {
+			let num_bss_nontx = mbssid_group_size - 1;
+			while (num_bss_nontx > 0) {
+				maxbssid_ind++;
+				num_bss_nontx >>= 1;
+			}
+		}
+		let maxbssid = 2 ** maxbssid_ind;
+		if (mbssid && idx >= maxbssid) {
+			addr[5] = ((addr[5] >> maxbssid_ind) << maxbssid_ind);
+			addr[5] = addr[5] + (idx / maxbssid) * maxbssid;
+			idx = idx % maxbssid;
+		}
+
 		let mask = macaddr_split(base_mask);
 		let type;
 
@@ -256,9 +272,15 @@ const phy_proto = {
 			addr[0] ^= idx << 2;
 			break;
 		case "b5":
-			if (mbssid)
+			if (mbssid) {
 				addr[0] |= 2;
-			addr[5] ^= idx;
+				let size = maxbssid;
+				let B = addr[5] % size;
+				let loct = addr[5];
+				addr[5] = loct - B + ((B+idx) % size);
+			} else {
+				addr[5] ^= idx;
+			}
 			break;
 		default:
 			for (let i = 5; i > 0; i--) {
-- 
2.34.1

